codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java

```diff
@@ -330,7 +331,10 @@ private static String decodeComponent(String s, int from, int toExcluded, Charse
 
             byteBuf.clear();
             do {
-                byteBuf.put(decodeHexByte(s, i, toExcluded));
+                if (i + 3 > toExcluded) {
+                    throw new IllegalArgumentException("unterminated escape sequence at index " + i + " of: " + s);
+                }
+                byteBuf.put(decodeHexDump(s, i + 1, 2));
```

Comment:
Oh I see what you are saying now ... makes sense. Not ideal to add a `ByteBuffer` method in `ByteBufUtil` ... maybe we should change `QueryStringDecoder` to use a `ByteBuf` (so we can take advantage of our allocators) and then use `ByteBuf#nioBuffer()` to do the decoding.